### PR TITLE
basic error returned from server

### PR DESCRIPTION
**Pull request by StephenCoady**
basic error returning from server with type. no client handling yet

### CODE REVIEW

File: server-apollo/index.js [hunk 1]
```diff
@@ -162,6 +181,10 @@ Mutation: {
 }
 }
 
+const conflictDetected = (server, client) => {
```
Comment (wtrocki) [inline server-apollo/index.js]: ping @darahayes 
We can have something like this method when conflict is detected.
Server side is just mock to showcase conflict resolution based on inmem data. We can effectively now showcase that conflict happens on the client and continue writing some handlers and other helpers that are needed.

**Comment by wtrocki** [comment]
@darahayes We have added basic versioning on client and server with logic that compares version.
This will allow us to continue working on the client without strict dependency on server side implementation you are doing. 

Take look into logic that is happening here:
- We use in-mem data
- Compare versions that were bumped on the client

I hope this will be useful.

File: web/src/apolloClient.js [hunk 2]
```diff
@@ -12,19 +13,34 @@ import { SyncOfflineMutation } from './mutations/SyncOfflineMutation'
...
 
+  const retryLink = new RetryLink({
+    delay: { initial: 5000, max: 10000 },
+    attempts: (count, operation, error) => {
+      console.log(operation)
+      console.log(error)
+        if (count > this.maxRetries) {
+            return false;
+        }
+        return false;
```
Comment (danielpassos) [inline web/src/apolloClient.js]: Is it always return false?

File: server-apollo/index.js [hunk 3]
```diff
@@ -1,5 +1,15 @@
 const { ApolloServer, gql } = require('apollo-server');
+const { GraphQLError } = require('graphql')
 
+class ConflictError extends GraphQLError {
+    constructor(type, data){
+        const message = "conflict has occured"
+        super(message)
+        this.data = data
+        this.type = type
```
Comment (wtrocki) [inline server-apollo/index.js]: If this is `ConflictError` then type makes no sense. Type can be used as string to check error on client.
I guess we may need just have something like `AGSyncError` or keep each error as separate class without type.
I have no strong opinion or preference.

@darahayes WDYT?

File: server-apollo/index.js [hunk 4]
```diff
@@ -99,43 +114,47 @@ Query: {
...
-    for (let user of users) {
-        if (args.id == user.id) {
-        user.name = args.name
-        user.feedback = args.feedback
-        user.dateOfBirth = args.dateOfBirth
-        return user
+        for (let user of users) {
+            if (args.id == user.id) {
+                if(conflictDetected(user.version, args.version)){
+                    throw new ConflictError('mutation', args)
```
Comment (wtrocki) [inline server-apollo/index.js]: What is 'mutation' in this context ?

File: server-apollo/index.js [hunk 5]
```diff
@@ -99,43 +114,47 @@ Query: {
...
     updateUser: (obj, args, context, info) => {
-    for (let user of users) {
-        if (args.id == user.id) {
-        user.name = args.name
-        user.feedback = args.feedback
-        user.dateOfBirth = args.dateOfBirth
-        return user
+        for (let user of users) {
+            if (args.id == user.id) {
+                if(conflictDetected(user.version, args.version)){
```
Comment (wtrocki) [inline server-apollo/index.js]: Let's throw some serious logging here so we could showcase this and see conflicts without debugging.

File: web/src/apolloClient.js [hunk 6]
```diff
@@ -12,19 +13,34 @@ import { SyncOfflineMutation } from './mutations/SyncOfflineMutation'
 
 export const setupApolloClient = async () => {
   const storage = window.localStorage
-  const uri = `https://api.graph.cool/simple/v1/cjmltohxn3phc0173w5w6p659`
+  const uri = `http://localhost:4000/graphql`
```
Comment (wtrocki) [inline web/src/apolloClient.js]:
```suggestion
  // Local server
  const uri = `http://localhost:4000/graphql`
  // Graph.cool for testing
  //const uri = `https://api.graph.cool/simple/v1/cjmltohxn3phc0173w5w6p659`
```

**Comment by danielpassos** [comment]
@StephenCoady I think the _version_ was miss in `AddUser`

```
writeToStore.js:154 Missing field version in {
  "__typename": "User",
  "id": "033UzjHa",
  "name": "User 1",
  "dateOfBirth": "Mon Oct 22 2018 
```

File: web/src/apolloClient.js [hunk 7]
```diff
@@ -12,19 +13,34 @@ import { SyncOfflineMutation } from './mutations/SyncOfflineMutation'
...
     console.log(response)
     response = { errors: null }
   })
 
+  const retryLink = new RetryLink({
+    delay: { initial: 5000, max: 10000 },
+    attempts: (count, operation, error) => {
+      console.log(operation)
+      console.log(error)
+        if (count > this.maxRetries) {
```
Comment (danielpassos) [inline web/src/apolloClient.js]:
```suggestion
        return count <= this.maxRetries;
```

File: web/src/apolloClient.js (hunk 2 above)
Comment (danielpassos) [reply web/src/apolloClient.js]: Maybe replace it by `return count <= this.maxRetries;`

File: web/src/view/ListUser.js [hunk 8]
```diff
@@ -52,7 +52,7 @@ class UserItem extends React.Component {
   onUpdate = async ({ item }) => {
     const { client } = this.props
     const { isOffline } = this.state
-    const variables = { name: `${item.name}_${Utils.generateId(1)}`, dateOfBirth: new moment(), id: item.id }
+    const variables = { name: `${item.name}_${Utils.generateId(1)}`, dateOfBirth: new moment(), id: item.id, version: item.version+1 }
```
Comment (wtrocki) [inline web/src/view/ListUser.js:55]: Maybe wrap this stuff into method so it will not be +1 everywhere..

**Comment by wtrocki** [comment]
@StephenCoady Optimistic response is missing versioning. Optimistic responses need to supply version (this is required parameter)

File: server-apollo/index.js [hunk 9]
```diff
@@ -99,43 +113,49 @@ Query: {
...
+                    // TODO add some logic to refetch if online
+                    console.log("CONFLICT DETECTED")
```
Comment (wtrocki) [inline server-apollo/index.js]: Can we log actual objects ? :>

**Comment by wtrocki** [comment]
@StephenCoady Rebased.. Going to made some small improvements in server. Feel free to experiment with server. Ideally we could merge that to master to improve collaboration and improve things there ;D

**Comment by darahayes** [comment]
@wtrocki @StephenCoady Nice PR! Overall I'm +1

**Review by danielpassos — APPROVED** [review]

**Pull request merged**